### DISCUSSION TITLE
Study Level Reporting - Data Pipeline

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#269] [Ideas]
# Data Pipeline 

Thinking more about standardizing and automating our data pipeline. Specifically thinking about data requirements and strategy for data checks (`stopifnot`) and tests for Map_Raw() and Assess(). 

# Standards for Map_Raw() Inputs

Map_Raw() should be flexible enough to support minor variations in input data. Proposing we implement the following standards: 

- dfRDSL is a required input for all map_raw functions. 
   - The expectation is that all `map_raw` functions will return a data frame with one record for each participant in RDSL. 
   - Manual merges with RDSL should be avoided. `util_merge_rdsl()` will be provided to merge `rdsl` with other data when needed. 
- All data columns and fields should be parameterized using the following naming conventions
    - dfYYY for data frames
    - strXXXCol for column names
    - vXXXVals for field values where XXX matches strXXXCol. For example, `strTrtCol="TRT"` and `strTrtVals = c("placebo", "active")`
    - Parameter names should be standardized across all Map_Raw, with a master list of mappings maintained in the data mapping vignette
- No NA, NaN or empty strings data allowed in any required column. 

# Standards for Assess() Input

Assess() should be rigid - dfInput will have specific column names for each assessment. (I think this is already true FWIW)
- dfInput should always have columns called SiteID and SubjectID.
- No NA, NaN or empty strings data allowed in any required column. Create a reusable `check_missing()` to confirm.    

# Data Checks

When possible, we should use standard functions to do our `stopifnot()` data checks: 
    - `check_columns()` confirms that expected columns are found
    - `check_values()` confirms that expected values are found 
    - `check_missing()` looks for missing values in required columns 
    - `check_type()` confirms that parameters have the expected type
    - `check_length()` confirms that parameters have the expected length 
We can consider consider whether this can all be wrapped up in to a single `check_params()` workflow. By moving these checks to their own functions, we can test them individually, and then remove the unit tests related to parameter checks for specific functions.

It's also possible (even likely) that we should just do all of this data/parameter checking using a dependency.

@kodesiba @mattroumaya @dsanders2gilead @gwu05 - definitely open to suggestions on implementation of these standardized data checks. 

# Issues List
- #159 
- Parameterize all columns and fields for map_raw (1 issue per assessment)
- Create full specs for the centralized data check process.

**Comment by jwildfire** [comment]
Note that this is all in support of simplifying running multiple assessments on a study. Will write more about that in a separate post.

**Comment by gwu05** [comment]
I think general idea sounds good. One specific issue is the mapping of RDSL to the outcome data required by the assessment which is more than one record per subject.

For example, for testing of binary, continuous, and/or survival endpoints, you'll often need to create dfInput with structure like:

columns: USUBJID, SITEID, PARAMCD, AVISITN, OUTCOME 

or for visit scheduling comparisons:

columns: USUBJID, SITEID, FDD, VisitDate, where visit date will be many per subject

Unfortunately, we won't be able to simply the outcomes into a single measure per subject for some of these analyses, as you'll need the visit info as well to do the analysis (for example, checking for duplicates at a visit). For testing of outcomes at a specific timepoint, we can simply by selecting a particular analysis value for that timepoint and require one value per subject.

**Reply by jwildfire** [reply]
Yeah, that makes sense. I'll remove the line about dfInput$SubjectID always being unique in the original post.

**Comment by jwildfire** [comment]
Options for external validation packages include:
- https://github.com/rich-iannone/pointblank
- https://github.com/ropensci/assertr/
- https://github.com/data-cleaning/validate
- https://github.com/ekstroem/dataReporter

**Comment by jwildfire** [comment]
Thinking we'll just write our own generalized data check function and avoid another dependency. Something like this: 

```
#' Check that a data frame contains columns and fields specified in mapping
#' 
#' @import dplyr
#'
#' @examples
#' rdsl_mapping<- list(id_col="SubjectID", site_col="SiteID", exposure_col="TimeOnTreatment")
#' is_mapping_valid(clindata::rdsl, rdsl_mapping, unique_cols="id_col") #TRUE
#' is_mapping_valid(clindata::rdsl, rdsl_mapping, unique_cols="site_col") #FALSE - site_col not unique
#' rdsl_mapping$not_a_col<-"nope"
#' is_mapping_valid(clindata::rdsl, rdsl_mapping, unique_cols="id_col") #FALSE - column not found
#'
#' @export
#' 
is_mapping_valid(df, mapping, unique_cols=NULL, na_cols=NULL){
    valid <- TRUE
    if(!is.data.frame(df)){
        message("df is not a data frame")
        valid<-FALSE
    }

    if(!nrows(df)>0){
        message("df has 0 rows")
        valid<-FALSE
    }

    # basic mapping checks
    if(!is.list(mapping)){
        message("Mapping is not a list")
        valid<-FALSE
    }

    if(!all(mapping%>%is.character(.x))){
        message("Non-characacter values found in mapping")
        valid<-FALSE
    }

    # expected columns not found in df
    expected_cols <- mapping[grep("_col", names(mapping))]
    expected_cols %>% map(function(col){
        if(!col %in% names(df)){
            message(paste0("Expected column not found: " ,col))  
            valid<-FALSE
        }
    })

    # No NA found in columns
    no_na_cols <- expected_cols[!names(expected_cols) %in% na_cols]
    no_na_cols %>% map(function(col){
        if(any(is.na(df[col]))){
            message(paste0("Unexpected NAs found in required column: ",col))
            valid<-FALSE
        }
        #TODO add NaN and blackspace checks
    })
    

    # Check that specified columns are unique
    if(!is.null(unique)){
        unique_cols <- expected_cols[names(expected_cols) %in% unique_cols]
        unique_cols %>% map(function(col){
            if(any(duplicated(df[[col]]))){
                message(paste0("Unexpected duplicates found column: ",col))
                valid<-FALSE
            }
        })
        
    }

    return(valid)
}
```

AE_Map_Raw gets refactored to something like this: 
```
AE_Map_Raw <- function( dfAE, dfRDSL, mapping = NULL )){
    if(is.null(mapping)){
        mapping <- list(
            dfAE= list(id_col="SubjectID"), 
            dfRDSL=list(strIDCol="SubjectID", strSiteCol="SiteID", strExposureCol="TimeOnTreatment")
        )
    }
    stopifnot(
        is_mapping_valid(dfAE,mapping$dfAE),
        is_mapping_valid(dfRDSL,mapping$RDSL) 
    )
   ...
}
```

With this framework, `mapping` and `dfXXX` become quite easy to share between assessments and study level data checking becomes pretty straightforward by running `is_mapping_valid` independent from the `map` functions.

**Reply by jwildfire** [reply]
Lots of details and refinements to be made in the above, but will get an initial PR together soon unless there are major concerns. @gwu05 @mattroumaya @kodesiba @dsanders2gilead

**Reply by jwildfire** [reply]
Made draft PR #273  with the code above to allow for more detailed discussion.

**Reply by jwildfire** [reply]
btw - I haven't actually run any of the code above, so we should all assume it is currently full of bugs :)

**Reply by mattroumaya** [reply]
Left a comment on the PR, but definitely like this approach! 🥇

**Comment by jwildfire** [comment]
Let's have `is_mapping_valid` return a list with the following properties:
- `status` - TRUE/FALSE
- `properties` - named list of boolean:
```
list(
  is_df=TRUE
  is_df_length_1 = TRUE,
  is_mapping_cols_char = TRUE,
  is_mapping_cols_found=FALSE,
...
),
```
- `messages` = char vector of all messages summarizing issues found

- Last line in function can be (Something like): `obj$status <- all(obj$properties)`
- Stop if not code becomes: 
``` 
stopifnot(
   is_mapping_valid(dfAE,mapping$dfAE)$status,
   is_mapping_valid(dfRDSL,mapping$RDSL)$status 
)
```
